src/Mod/PartDesign/Gui/TaskPocketParameters.cpp

```diff
@@ -291,6 +353,193 @@ void TaskPocketParameters::onOffsetChanged(double len)
+    int indexOfCurrent = -1;
+    App::DocumentObject* ax = propReferenceAxis->getValue();
+    const std::vector<std::string>& subList = propReferenceAxis->getSubValues();
+    for (size_t i = 0; i < axesInList.size(); i++) {
+        if (ax == axesInList[i]->getValue() && subList == axesInList[i]->getSubValues()) {
+            indexOfCurrent = i;
+            break;
+        }
+    }
+    // if the axis is not yet listed in the combobox
+    if (indexOfCurrent == -1 && ax) {
+        assert(subList.size() <= 1);
+        std::string sub;
+        if (!subList.empty())
+            sub = subList[0];
+        addAxisToCombo(ax, sub, getRefStr(ax, subList));
+        indexOfCurrent = axesInList.size() - 1;
+        // the axis is not the normal, thus enable along direction
+        ui->checkBoxAlongDirection->setEnabled(true);
+        // we don't have custom direction thus disable its settings
+        ui->XDirectionEdit->setEnabled(false);
+        ui->YDirectionEdit->setEnabled(false);
+        ui->ZDirectionEdit->setEnabled(false);
+    }
+
+    // highlight either current index or set custom direction
+    PartDesign::Pocket* pcPocket = static_cast<PartDesign::Pocket*>(vp->getObject());
+    bool hasCustom = pcPocket->UseCustomVector.getValue();
+    if (indexOfCurrent != -1 && !hasCustom)
+        ui->directionCB->setCurrentIndex(indexOfCurrent);
+    if (hasCustom)
+        ui->directionCB->setCurrentIndex(2);
+
+    blockUpdate = oldVal_blockUpdate;
+}
+
+void TaskPocketParameters::addAxisToCombo(App::DocumentObject* linkObj,
+    std::string linkSubname, QString itemText)
+{
+    this->ui->directionCB->addItem(itemText);
+    this->axesInList.emplace_back(new App::PropertyLinkSub);
+    App::PropertyLinkSub& lnk = *(axesInList.back());
+    lnk.setValue(linkObj, std::vector<std::string>(1, linkSubname));
+}
+
+void TaskPocketParameters::onDirectionCBChanged(int num)
+{
+    PartDesign::Pocket* pcPocket = static_cast<PartDesign::Pocket*>(vp->getObject());
+
+    if (axesInList.empty() || !pcPocket)
+        return;
+
+    App::PropertyLinkSub& lnk = *(axesInList[num]);
+    if (lnk.getValue() == 0) {
+        // enter reference selection mode
+        this->blockConnection(false);
+        // to distinguish that this is the direction selection
+        selectionFace = false;
+        TaskSketchBasedParameters::onSelectReference(true, true, false, true, true);
+        return;
+    }
+    else {
+        if (!pcPocket->getDocument()->isIn(lnk.getValue())) {
+            Base::Console().Error("Object was deleted\n");
+            return;
+        }
+        propReferenceAxis->Paste(lnk);
+        // in case user is in selection mode, but changed his mind before selecting anything
+        exitSelectionMode();
+    }
+
+    try {
+        recomputeFeature();
+    }
+    catch (const Base::Exception& e) {
+        e.ReportException();
+    }
+
+    // disable AlongSketchNormal when the direction is already normal
+    if (num == 0)

```

Comment:
Is there a way to refactor to use an `enum` for these `num` comparisons, or even just some named constants, instead of bare numbers here?

Reply:
The `num` is an int because I number the possible settings according to the combobox entries:
0 - sketch normal
1 - entry to select a reference
2 - custom direction
3, 4, 5... - the different edges the user had selected in the 3D model

For the latter I refer to the int to get the particular axis. Since the list of edges has no limit, I cannot use an enum.
Do you have another idea?

Reply:
You could still use an enum if you wanted to, it's not really much different than defining three independent constants. Maybe something like:
```
enum class ComboboxMapping {
  SKETCH_NORMAL,
  SELECTED_REFERENCE,
  CUSTOM_DIRECTION,
  SELECTED_EDGES_OFFSET // Could omit this one if there's no use for it -- do you ever add/subtract 3 to get the index?
};
```
You could also just:
```
const int SKETCH_NORMAL = 0;
const int SELECTED_REFERENCE = 1;
// etc.
```

Reply:
I don't understand yet. What I do is tho read the current combobox entry. I take the entry index as number and this is the also the index number for the axes list

> do you ever add/subtract 3 to get the index?

No. Since I return the current combobox index, I already have this index number ans the index number of the axes list.

So assuming there are 4 axes from the 3D model, then the combo has 7 entries. When the user sets the combo to index number 5, I know this is the axis axesInList[5], see line 429.

Reply:
Maybe a more concrete example: I'd rather see code like this:
```
if (num == SKETCH_NORMAL)
```
or this:
```
if (num == ComboboxMapping::SKETCH_NORMAL)
```
than this
```
if (num == 0)
```
It's not critical, it just makes it a little easier to read and reason about, and is self-documenting.